src/PrestaShopBundle/Controller/Admin/Improve/Payment/PaymentPreferencesController.php

```diff
@@ -89,21 +100,92 @@ public function indexAction(Request $request)
+        return $this->processForm($carrierRestrictionsFormHandler, $request);
+    }
 
-        if ($paymentPreferencesForm->isSubmitted()) {
-            $paymentPreferences = $paymentPreferencesForm->getData();
+    /**
+     * Process payment modules preferences country restrictions form.
+     *
+     * @AdminSecurity(
+     *     "is_granted(['update', 'create', 'delete'], request.get('_legacy_controller'))",
+     *     message="Access denied.",
+     *     redirectRoute="admin_payment_preferences"
+     * )
+     *
+     * @param Request $request
+     *
+     * @return RedirectResponse
+     */
+    public function processPaymentCountryRestrictionsFormAction(Request $request): Response
+    {
+        $countryRestrictionsFormHandler = $this->getPaymentCountryRestrictionsFormHandler();
 
-            $errors = $paymentPreferencesFormHandler->save($paymentPreferences);
-            if (empty($errors)) {
-                $this->addFlash('success', $this->trans('Successful update.', 'Admin.Notifications.Success'));
+        return $this->processForm($countryRestrictionsFormHandler, $request);
+    }
 
-                return $this->redirectToRoute('admin_payment_preferences');
+    /**
+     * Process payment modules preferences currency restrictions form.
+     *
+     * @AdminSecurity(
+     *     "is_granted(['update', 'create', 'delete'], request.get('_legacy_controller'))",
+     *     message="Access denied.",
+     *     redirectRoute="admin_payment_preferences"
+     * )
+     *
+     * @param Request $request
+     *
+     * @return RedirectResponse
+     */
+    public function processPaymentCurrencyRestrictionsFormAction(Request $request): Response
+    {
+        $currencyRestrictionsFormHandler = $this->getPaymentCurrencyRestrictionsFormHandler();
+
+        return $this->processForm($currencyRestrictionsFormHandler, $request);
+    }
+
+    /**
+     * Process payment modules preferences group restrictions form.
+     *
+     * @AdminSecurity(
+     *     "is_granted(['update', 'create', 'delete'], request.get('_legacy_controller'))",
+     *     message="Access denied.",
+     *     redirectRoute="admin_payment_preferences"
+     * )
+     *
+     * @param Request $request
+     *
+     * @return RedirectResponse
+     */
+    public function processPaymentGroupRestrictionsFormAction(Request $request): Response
+    {
+        $groupRestrictionsFormHandler = $this->getPaymentGroupRestrictionsFormHandler();
+
+        return $this->processForm($groupRestrictionsFormHandler, $request);
+    }
+
+    /**
+     * Processes the form in a generic way.
+     *
+     * @param FormHandlerInterface $formHandler
+     * @param Request $request
+     *
+     * @return RedirectResponse
+     */
+    private function processForm(FormHandlerInterface $formHandler, Request $request): RedirectResponse
+    {
+        $form = $formHandler->getForm();
+        $form->handleRequest($request);
+
+        if ($form->isSubmitted()) {
```

Comment:
Shouldn't we check if form is valid?

Reply:
Because we redirecting back to original url constraints won't work. This makes checking for is valid pointless. Even if there would be errors raised by constraints making form invalid, user won't see them. So it's better to validate data by dataProvider in this case. If there is anything to validate.

Reply:
there are no constraints in form as I see for now, but such errors can be handled by `forwarding` the request instead of `redirecting` https://symfony.com/doc/4.4/controller/forwarding.html. (idk if that's a good practice, but i see it to be the only option if we want to have separate actions for form handling). I've quickly tested it and form errors seem to appear at debug toolbar this way, though, you need to pass the previous request params (so at least the legacy_controller, because else you wont be able to access index page)
With your current implementation if some module wants to extend the form with custom field, constraints will not be handled, so it doesn't seem right. And `validate data by dataProvider` sounds even worse

Reply:
So if a module override the form and add a contraint, it won't be checked. Isn't it ?

Reply:
> So if a module override the form and add a constraint, it won't be checked. Isn't it ?

Now it won't. I will need to check maybe the forwarding thing Julius mentioned would be a solution.

Reply:
Regarding forward I tried testing it and it does not seem to solve the issue. 

If we have multiple forms we need separate actions(and urls for those actions) for saving, we can't just save all forms in same index action. After saving we can't stay at /country-restrictions and using forward does exactly that.

We can't stay at /country-restrictions because that route is only configured only for post request, so if user manages to refresh page in some way (like pagination) he will get a crash.

Reply:
yeah well, I also tried some possible solutions, but without success :/
I didn't find any useful examples in symfony docs either as it assumes one form per action everywhere or suggests using collections.
Closest one I tried was submitting all of them to same action and checking which one was submitted. Though it still ended up a failure (there was the violation mesage in profiler as expected, but I failed to render it probably because of its wrong property path. It may be due to fact that there are 4 forms rendered, but only 1 submitted, so then the violation path seems wrong).
Still, the most obvious indicator that something smells here is that form constraints becomes useless in current implementation :sweat: Maybe its acceptable? but we should warn module developers then.
maybe @jolelievre has any idea about it? :thinking: